pkg/monitors/filesystems/partitions_windows.go

```diff
@@ -0,0 +1,195 @@
+		if err != nil {
+			return stats, errors.WithMessagef(err, "cannot find partition-stats for first-volume %s", windows.UTF16ToString(volNameBuf))
+		}
+		stats = append(stats, partitionStats...)
+	}
+
+	var lastError error
+	for {
+		volNameBuf = make([]uint16, volumeNameBufferLength)
+		if err = findNextVolume(handle, &volNameBuf[0]); err != nil {
+			if errno, ok := err.(syscall.Errno); ok && errno == windows.ERROR_NO_MORE_FILES {
+				break
+			}
+			lastError = errors.WithMessagef(err, "find next-volume last-error")
+			continue
+		}
+
+		volPaths, err = getVolumePaths(volNameBuf)
+		if err != nil {
+			lastError = errors.WithMessagef(err, "find paths last-error for volume %s", windows.UTF16ToString(volNameBuf))
+			continue
+		}
+
+		if len(volPaths) > 0 {
+			partitionStats, err = getPartitionStats(getDriveType(volPaths[0]), volPaths, getFsNameAndFlags)
+			if err != nil {
+				lastError = errors.WithMessagef(err, "find partition-stats last-error for volume %s", windows.UTF16ToString(volNameBuf))
+				continue
+			}
+			stats = append(stats, partitionStats...)
+		}
+	}
+
+	return stats, lastError
+}
+
+func getDriveType(rootPath string) (driveType uint32) {
+	rootPathPtr, _ := windows.UTF16PtrFromString(rootPath)
+	return windows.GetDriveType(rootPathPtr)
+}
+
+func findFirstVolume(volName *uint16) (findVol windows.Handle, err error) {
+	return windows.FindFirstVolume(volName, volumeNameBufferLength)
+}
+
+func findNextVolume(findVol windows.Handle, volName *uint16) (err error) {
+	return windows.FindNextVolume(findVol, volName, volumeNameBufferLength)
+}
+
+func findVolumeClose(findVol windows.Handle) (err error) {
+	return windows.FindVolumeClose(findVol)
+}
+
+// getVolumePaths returns the path for the given volume name.
+func getVolumePaths(volNameBuf []uint16) ([]string, error) {
+	volPathsBuf := make([]uint16, volumePathBufferLength)
+	returnLen := uint32(0)
+	if err := windows.GetVolumePathNamesForVolumeName(&volNameBuf[0], &volPathsBuf[0], volumePathBufferLength, &returnLen); err != nil {
+		return nil, err
+	}
+
+	return findStrings(volPathsBuf, int(returnLen)), nil
+}
+
+// findStrings returns strings from subslices in uint16s delimited by 0.
+func findStrings(uint16s []uint16, stopAt int) []string {
+	if stopAt > len(uint16s) {
+		stopAt = len(uint16s)
+	}
+
+	from, ss := 0, make([]string, 0)
+
+	for to := 0; to < stopAt; to++ {
+		if uint16s[to] == 0 {
+			if from < to && uint16s[from] != 0 {
+				ss = append(ss, string(utf16.Decode(uint16s[from:to])))
+			}
+			from = to + 1
+		}
+	}
+	return ss
+}
+
+// getFsNameAndFlags sets inputs fsNameBuf and fsFlags with fetched filesystem name (NTFS, FAT32 etc) and flags.
+func getFsNameAndFlags(rootPath string, fsNameBuf []uint16, fsFlags *uint32) error {
+	volNameBuf := make([]uint16, 256)
+	volSerialNum := uint32(0)
+	maxComponentLen := uint32(0)
+	rootPathPtr, _ := windows.UTF16PtrFromString(rootPath)
```

Comment:
same hidden error concern

Reply:
rootPath never contains null bytes, so error not possible